hoot-core/src/main/cpp/hoot/core/algorithms/ReplacementSnappedWayJoiner.cpp

```diff
@@ -0,0 +1,151 @@
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2019 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "ReplacementSnappedWayJoiner.h"
+
+// Hoot
+#include <hoot/core/util/Factory.h>
+#include <hoot/core/ops/ReplaceElementOp.h>
+#include <hoot/core/io/OsmMapWriterFactory.h>
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(WayJoiner, ReplacementSnappedWayJoiner)
+
+ReplacementSnappedWayJoiner::ReplacementSnappedWayJoiner()
+{
+  _leavePid = true;
+  _callingClass = QString::fromStdString(className());
+}
+
+ReplacementSnappedWayJoiner::ReplacementSnappedWayJoiner(
+  const QMap<ElementId, long>& refIdToVersionMappings) :
+WayJoinerAdvanced::WayJoinerAdvanced(),
+_refIdToVersionMappings(refIdToVersionMappings)
+{
+  _leavePid = true;
+  _callingClass = QString::fromStdString(className());
+}
+
+bool ReplacementSnappedWayJoiner::_areJoinable(const WayPtr& w1, const WayPtr& w2) const
+{
+  // We'll join up anything, as long as its not invalid.
+  return w1->getStatus() != Status::Invalid && w2->getStatus() != Status::Invalid;
+}
+
+void ReplacementSnappedWayJoiner::_determineKeeperFeatureForTags(
+  WayPtr parent, WayPtr child, WayPtr& keeper, WayPtr& toRemove) const
+{
+  // We always want to keep unknown2, which is the dough ways being joined up with the cookie cut
+  // replacement ways.
+  if (parent->getStatus() == Status::Unknown2)
+  {
+    keeper = parent;
+    toRemove = child;
+  }
+  else if (child->getStatus() == Status::Unknown2)
+  {
+    keeper = child;
+    toRemove = parent;
+  }
+  else
+  {
+    keeper = parent;
+    toRemove = child;
+  }
+}
+
+void ReplacementSnappedWayJoiner::_determineKeeperFeatureForId(
+  WayPtr parent, WayPtr child, WayPtr& keeper, WayPtr& toRemove) const
+{
+  _determineKeeperFeatureForTags(parent, child, keeper, toRemove);
+}
+
+bool ReplacementSnappedWayJoiner::_hasPid(const ConstWayPtr& way) const
+{
+  return way->hasPid() || way->getTags().contains(MetadataTags::HootSplitParentId());
+}
+
+long ReplacementSnappedWayJoiner::_getPid(const ConstWayPtr& way) const
+{
+  long pid = WayData::PID_EMPTY;
+  if (way->hasPid())
+  {
+    pid = way->getPid();
+  }
+  else if (way->getTags().contains(MetadataTags::HootSplitParentId()))
+  {
+    pid = way->getTags()[MetadataTags::HootSplitParentId()].toLong();
+  }
+  return pid;
+}
+
+void ReplacementSnappedWayJoiner::join(const OsmMapPtr& map)
+{
+  LOG_DEBUG("Joining ways...");
+
+  WayJoinerAdvanced::join(map);
+
+  if (_refIdToVersionMappings.isEmpty())
+  {
+    LOG_WARN("No version mappings exist for ref ways.");
+  }
+
+  // If anything left has a PID on it, let's make that PID its ID. This doesn't seem like a great
+  // way to do it and problems could be encountered in the future where more than one split way inside
+  // of the bounds had the same parent...haven't seen that happen yet, so will go with this for now.
+  WayMap ways = _map->getWays();
```

Comment:
Out of everything in this PR, this is probably what I'm least proud of and possibly likely to break as we test with more data. Its primarily focused around making sure that ref ID's are preserved throughout the changeset generation process and that you end up with changeset modify statements for existing features. This code mostly is for the linear feature ID provenance. Anyway, we'll see how long it holds up...works well with the roads in the Boston test dataset I've been using.

Reply:
ID preservation is one of the hardest things.  This isn't a bad idea, the way came from somewhere and if the parent doesn't exist then using that ID definitely preserves it.